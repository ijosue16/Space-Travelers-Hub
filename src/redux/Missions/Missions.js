import { createAsyncThunk, createAction, createReducer } from '@reduxjs/toolkit';

const GET_MISSIONS = 'space traveler hub/missions/GET_MISSIONS';
const URL = 'https://api.spacexdata.com/v3/missions';
const INITIALSTATE = [];

const getMissionsAPI = async () => {
  const resp = await fetch(URL);
  const data = await resp.json();
  return data;
};
export const getMissions = createAsyncThunk(GET_MISSIONS, async (e, thunkAPI) => {
  const currentState = thunkAPI.getState();
  if (currentState.Missions.length === 0) {
    const resp = await getMissionsAPI();
    const missionsData = [];
    resp.forEach((mission) => {
      const singleMission = {
        id: mission.mission_id,
        name: mission.mission_name,
        description: mission.description,
        reserved: false,
      };
      missionsData.push(singleMission);
    });
    return missionsData;
  }
  return currentState.Missions;
});

export const updateMission = createAction('mission/UPDATE_MISSION');

const Missionreducer = createReducer(INITIALSTATE, ((builder) => {
  builder
    .addCase(getMissions.fulfilled, ((state, action) => action.payload))
    .addCase(updateMission, (state, { payload }) => state.map((mission) => {
      if (mission.id === payload) {
        return { ...mission, reserved: !mission.reserved };
      }
      return mission;
    }))
    .addDefaultCase();
}));

export default Missionreducer;
