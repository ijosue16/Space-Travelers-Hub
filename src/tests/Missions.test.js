import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import store from '../redux/configurestore';
import Missions from '../pages/missionslist';
import '@testing-library/jest-dom';

it('missionslist renders correctly', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Missions />
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

describe('mission test', () => {
  test('check if mission renders empty initially ', () => {
    render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );
    expect(screen.queryByText('chaicom')).not.toBeInTheDocument();
  });
  it('missionslist renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Missions />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
