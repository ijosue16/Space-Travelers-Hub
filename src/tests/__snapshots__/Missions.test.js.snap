// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mission test missionslist renders correctly 1`] = `
<div
  className="container-fluid d-flex gap-2 flex-column "
>
  <table
    className="table table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
        <th>
           
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;

exports[`missionslist renders correctly 1`] = `
<div
  className="container-fluid d-flex gap-2 flex-column "
>
  <table
    className="table table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
        <th>
           
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
